### PR TITLE
Global crypto provider

### PR DESCRIPTION
**Pull request by Valodim**
This PR moves the crypto provider setting and the "support unsigned" one from account settings to global.

This is a working state, but it's still an intermediate PR:
- there is now a "Cryptography" setting in both account and global settings, which is confusing
- there is no migration, so users have to set the global crypto provider setting anew after the update

To make things a little clearer, a toast is displayed when the account setting is clicked while no crypto provider is configured that informs the user to check their global settings.

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/preferences/Settings.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
      *
      * @see SettingsExporter
      */
-    public static final int VERSION = 44;
+    public static final int VERSION = 46;
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/preferences/Settings.java]: (the reason this is incremented by 2 is because the settings were moved in separate commits)

File: k9mail/src/main/java/com/fsck/k9/preferences/AccountSettings.java [hunk 2]
```diff
@@ -53,10 +53,6 @@
         s.put("chipColor", Settings.versions(
                 new V(1, new ColorSetting(0xFF0000FF))
             ));
-        s.put("cryptoApp", Settings.versions(
-                new V(1, new StringSetting("apg")),
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/preferences/AccountSettings.java]: Does setting this to null remove the setting? I think it does, but this is never done anywhere and the code that deals with this isn't very simple :\

**Comment by Valodim** [comment]
I consider this PR done. The above mentioned issues will be solved in follow-up PRs:
- the "Cryptography" settings in account will be removed when the setting is moved into identities
- migration will be sidestepped in a PR that adds a workflow for configuring crypto provider when an encrypted message is viewed while none is configured

File: k9mail/src/main/java/com/fsck/k9/K9.java [hunk 3]
```diff
@@ -465,6 +469,9 @@ public static void save(StorageEditor editor) {
         editor.putBoolean("hideUserAgent", mHideUserAgent);
         editor.putBoolean("hideTimeZone", mHideTimeZone);
 
+        editor.putString("cryptoProvider", sCryptoProvider);
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: I would really prefer if this sort of setting referred to OpenPGP rather than 'Crypto'.

File: k9mail/src/main/java/com/fsck/k9/preferences/Settings.java [hunk 4]
```diff
@@ -36,7 +36,7 @@
      *
      * @see SettingsExporter
      */
-    public static final int VERSION = 45;
+    public static final int VERSION = 47;
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/preferences/Settings.java]: 46?

File: k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSettings.java [hunk 5]
```diff
@@ -739,11 +710,21 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
                 }
             });
 
-            mCryptoSupportSignOnly.setChecked(mAccount.getCryptoSupportSignOnly());
+            cryptoMenu.setOnPreferenceClickListener(null);
         } else {
-            final Preference mCryptoMenu = findPreference(PREFERENCE_CRYPTO);
-            mCryptoMenu.setEnabled(false);
-            mCryptoMenu.setSummary(R.string.account_settings_no_openpgp_provider_installed);
+            cryptoMenu.setSummary(R.string.account_settings_no_openpgp_provider_configured);
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/setup/AccountSettings.java]: See this explicitly binds 'crypto' to meaning OpenPGP. It's not very forward compatible.

File: k9mail/src/main/java/com/fsck/k9/preferences/Settings.java (hunk 4 above)
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/preferences/Settings.java]: I moved the two settings independently in two commits, hence the two version steps you see in the full diff.

File: k9mail/src/main/java/com/fsck/k9/K9.java (hunk 3 above)
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: What's to say the provider can't also provide other crypto? So far I put OpenPGP only in places where actual openpgp things happened, and use "crypto" where it's mostly generic. See `CryptoError` and such. Right now I'd say it's more likely that OpenKeychain grows S/MIME support than someone coming along with an api of their own for this.

Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: I see the S/MIME API being similar but I don't think it needs to be the same app. In fact I don't think it should be. Otherwise every app that uses the API has to support both.

We should have a PGP API and an S/MIME API. They are fundamentally different and so should be different APIs. If OpenKeychain wanted to implement both that would be a decision for that app.

So this should talk about pgpProvider.

**Comment by cketti** [comment]
Selecting "<none>" as key in account settings doesn't seem to have any effect. The crypto UI is still displayed for this account. And messages encrypted for the previously selected key are still decrypted.

**Comment by Valodim** [comment]
the key is only selected for signing, decryption works for any key that k9 is allowed to use. that behavior didn't change in this PR, but it's one of the things that will change in upcoming PRs.

File: k9mail/src/main/java/com/fsck/k9/K9.java (hunk 3 above)
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: renamed

File: plugins/openpgp-api-lib/openpgp-api/src/main/java/org/openintents/openpgp/util/OpenPgpAppPreference.java [hunk 6]
```diff
@@ -270,19 +270,21 @@ private void populateAppList() {
...
-
-            String packageName = resolveInfo.serviceInfo.packageName;
-            String simpleName = String.valueOf(resolveInfo.serviceInfo.loadLabel(getContext()
-                    .getPackageManager()));
-            Drawable icon = resolveInfo.serviceInfo.loadIcon(getContext().getPackageManager());
-
-            if (!PROVIDER_BLACKLIST.contains(packageName)) {
-                mList.add(new OpenPgpProviderEntry(packageName, simpleName, icon));
-                hasNonBlacklistedChoices = true;
+        if (resInfo != null) {
```
Comment (dschuermann) [inline plugins/openpgp-api-lib/openpgp-api/src/main/java/org/openintents/openpgp/util/OpenPgpAppPreference.java]: please backport changes to the openpgp-api to our git.

File: k9mail/src/main/java/com/fsck/k9/K9.java (hunk 3 above)
Comment (dschuermann) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: @Valodim Hmm, this is still "cryptoProvider"?

Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/K9.java]: forgot a few

File: k9mail/src/main/res/xml/global_preferences.xml [hunk 7]
```diff
@@ -415,12 +415,12 @@
 
         <org.openintents.openpgp.util.OpenPgpAppPreference
             android:persistent="false"
-            android:key="crypto_app"
+            android:key="openpgp_provider"
```
Comment (dschuermann) [inline k9mail/src/main/res/xml/global_preferences.xml]: shouldn't this be "openpgpProvider" in accordance with `GlobalSettings``?

File: k9mail/src/main/res/xml/global_preferences.xml [hunk 8]
```diff
@@ -415,12 +415,12 @@
...
             android:title="@string/account_settings_crypto_app" />
 
         <CheckBoxPreference
             android:persistent="false"
-            android:key="crypto_support_sign_only"
+            android:key="openpgp_support_sign_only"
```
Comment (dschuermann) [inline k9mail/src/main/res/xml/global_preferences.xml]: and "openpgpSupportSignOnly"?

File: k9mail/src/main/res/xml/global_preferences.xml (hunk 7 above)
Comment (Valodim) [inline k9mail/src/main/res/xml/global_preferences.xml]: corresponds to `PREFERENCE_OPENPGP_PROVIDER` in `Prefs`

Comment (dschuermann) [inline k9mail/src/main/res/xml/global_preferences.xml]: But the upgrade procedure in ``GlobalSettings`` works by the key or am I wrong? I mean you could also update the key in ``GlobalSettings`` to "openpgp_provider", but it needs to be the same key, right?

As an example: "fixedMessageViewTheme" is written the same in GlobalSettings, Prefs and global_preferences.

Comment (Valodim) [inline k9mail/src/main/res/xml/global_preferences.xml]: We don't use a SharedPreferences to back this but do the saving manually in `saveSettings`, so the key is independent. We *could* go for consistency and name it the same everywhere, but that also masks how these values are really from different namespaces, leading to wrong assumptions about the code structure.

Comment (dschuermann) [inline k9mail/src/main/res/xml/global_preferences.xml]: Hmmm okay. Isn't this then true for all keys inside the xml files? At least persistent is always "false". Maybe the naming should be unified in another PR if this is the case. Currently it is confusing.

Comment (dschuermann) [inline k9mail/src/main/res/xml/global_preferences.xml]: I added a TODO: https://github.com/k9mail/k-9/issues/2213

**Comment by dschuermann** [comment]
LGTM 👍
